### PR TITLE
Migration to 3.4

### PR DESCRIPTION
**Pull request by priyankarani**

### CODE REVIEW

File: sale.py [hunk 1]
```diff
@@ -226,3 +245,120 @@ def process(cls, sales):
...
+    __name__ = 'sale.payment'
+
+    gift_card = fields.Many2One(
+        "gift_card.gift_card", "Gift Card", states={
+            'required': Eval('method') == 'gift_card',
+            'invisible': ~(Eval('method') == 'gift_card'),
+        }, domain=[('state', '=', 'active')], depends=['method']
+    )
+
+    def _create_payment_transaction(self, amount, description):
```
Comment (prakashpp) [inline sale.py]: why this?

File: sale.py [hunk 2]
```diff
@@ -226,3 +245,120 @@ def process(cls, sales):
...
+            )
+
+
+class AddSalePayment(Wizard):
+    """
+    Wizard to add a Sale Payment
+    """
+    __name__ = 'sale.payment.add'
+
+    def create_sale_payment(self, profile=None):
```
Comment (prakashpp) [inline sale.py]: why are you reimplementing it?

File: sale.py [hunk 3]
```diff
@@ -206,6 +210,14 @@ class Sale:
     "Sale"
     __name__ = 'sale.sale'
 
+    @staticmethod
+    def default_payment_authorize_on():
```
Comment (sharoonthomas) [inline sale.py]: why should gift card (just another gateway) module decide the defaults on how the payment should be captured or authorized. Please remove.

File: sale.py [hunk 4]
```diff
@@ -206,6 +210,14 @@ class Sale:
...
+        return 'manual'
+
+    @staticmethod
+    def default_payment_capture_on():
```
Comment (sharoonthomas) [inline sale.py]: remove

File: sale.py [hunk 5]
```diff
@@ -214,6 +226,13 @@ def create_gift_cards(self):
             line.create_gift_cards()
 
     @classmethod
+    def get_payment_method_priority(cls):
+        """Priority order for payment methods. Downstream modules can override
+        this method to change the method priority.
+        """
+        return ('gift_card', 'manual', 'credit_card')
```
Comment (sharoonthomas) [inline sale.py]: `return ('gift_card',) + super(...)`

File: sale.py (hunk 1 above)
Comment (sharoonthomas) [inline sale.py]: This looks wrong. If the payment gateway's method design has to change, it has to, but overwriting so much is not a good idea.

File: sale.py [hunk 6]
```diff
@@ -226,3 +245,120 @@ def process(cls, sales):
...
+        """
+        Validate sale payment to add gift card with sufficient amount only
+        """
+        super(Payment, cls).validate(payments)
+
+        for payment in payments:
+            if payment.gift_card:
+                payment.check_gift_card_amount()
+
+    def check_gift_card_amount(self):
```
Comment (sharoonthomas) [inline sale.py]: this validation would be triggered on all edits to `sale.payment` even after the payment is consumed. In the consumed state, the gift_card will never have a balance ? and this validation would always fail.

PaymentTransaction has the logic to check for gift card balances on authorization and capture and there is no reason why this should check here as well. OTOH, displaying available balance is an API choice.

**Pull request merged**